### ISSUE TITLE
multiple issues when upgrading from Capybara 2.x to 3.15.0

### ISSUE THREAD

**Issue by monfresh** [teamcapybara/capybara#2170]
## Meta
Capybara Version:
3.15.0
Driver Information (and browser if relevant):
selenium-webdriver 3.141.0 with chromedriver 2.45

## Expected Behavior
After upgrading to Capybara 3.x, tests continue to pass

## Actual Behavior
Certain tests that involve scrolling the page and checking for Rails flash messages don't pass

## Steps to reproduce
On our master branch, we are using Capybara version 2.18.0 with chromedriver 2.45, and we are not having any issues. https://github.com/department-of-veterans-affairs/caseflow/blob/master/spec/support/capybara.rb

After upgrading to Capybara 3.15.0, which was the only change, some of our tests are failing. One of them involves scrolling a document and checking to see that the page number changes when you scroll to the second page. This works fine in Capybara 2.18 but not in 3.15.

Another test that is failing checks for a flash message on the page. When I run the test using Chrome with Capybara 2.18, I see the flash message, but with 3.15, I don't see it.

Any idea what might be going on in these 2 types of tests? This is a Rails app with a React frontend. Again, I'm testing this on a branch where the only difference is the new version of Capybara.

Thanks for any insight you can provide!

**Comment by monfresh** [comment]
Aha. It looks like it's related to using puma as the Capybara server. We use puma in our application, so we left it as is (per the upgrade guide). I just tried setting the server to webrick, and now the tests pass! Have you heard of issues like this with puma?

**Comment by twalpole** [comment]
Please give an example of a test that fails and the complete error message you get

**Comment by monfresh** [comment]
Here is an example that fails only with Capybara 3.x, and it fails with both webrick and puma:

https://github.com/department-of-veterans-affairs/caseflow/blob/master/spec/feature/reader/reader_spec.rb#L773-L784

```ruby
scenario "Scrolling pages changes page numbers" do
  visit "/reader/appeal/#{appeal.vacols_id}/documents"

  click_on documents[1].type

  expect(page).to have_content("IN THE APPEAL", wait: 10)

  expect(page).to have_css(".page")
  expect(find_field("page-progress-indicator-input").value).to eq "1"
  scroll_to(class_name: "ReactVirtualized__Grid", value: 2000, index: 1)
  expect(find_field("page-progress-indicator-input").value).to_not eq "1"
end
```
Here is the definition of `scroll_to`:
```ruby
def scroll_to(id: nil, class_name: nil, value: 0, index: 0)
  page.driver.evaluate_script <<-EOS
    function() {
      var elem = document.getElementById('#{id}') || document.getElementsByClassName('#{class_name}')['#{index}'];
      elem.scrollTop=#{value};
    }();
  EOS
end
```

The test fails with this message:
```
Failure/Error: expect(find_field("page-progress-indicator-input").value).to_not eq "1"

       expected: value != "1"
            got: "1"

       (compared using ==)
```

When I run the test in Chrome with Capybara 2.18 and insert a sleep after `scroll_to(class_name: "ReactVirtualized__Grid", value: 2000, index: 1)`, I don't see the page scrolling, but I do see the page number at the bottom change. When I run the same test in 3.15, I don't see the page number change.

**Comment by twalpole** [comment]
You're writing your expectations using RSpecs generic matchers (`eq`) rather than the Capybara provided matchers which handle the asynchronous nature of JS driven pages.  Rather than `expect(find_field("page-progress-indicator-input").value).to eq "1"` you should be writing

    expect(page).to have_field("page-progress-indicator-input", with: "1")

and then either 

    expect(page).not_to have_field("page-progress-indicator-input", with: "1")

or

    expect(find_field("page-progress-indicator-input")).not_to match_selector(:field, with: "1")

depending on exactly what you're checking for (truthfully you're generally better off checking for what the new value is expected to be rather than just that it is no longer the old value).  Also note that Capybara has scrolling support built-in now https://www.rubydoc.info/gems/capybara/Capybara/Node/Element#scroll_to-instance_method so your scroll_to method may not be needed any longer

**Comment by monfresh** [comment]
Thank you very much! That works. It's funny, after I posted this issue, I came across a [thoughtbot article](https://thoughtbot.com/blog/write-reliable-asynchronous-integration-tests-with-capybara) explaining why one should use the matchers instead of calling methods on nodes, and I did not make the connection with this particular spec. The built-in `scroll_to` works great as well. I ended up doing this:
```ruby
all(".ReactVirtualized__Grid").last.scroll_to(0, 2000)

expect(page).to_not have_field("page-progress-indicator-input", with: "1")
```
